metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala

```diff
@@ -254,22 +265,66 @@ class DestinationProvider(
     }
   }
 
-  def fromSymbol(symbol: String): Option[DefinitionDestination] = {
-    for {
-      symbolDefinition <- index.definition(Symbol(symbol))
-      if symbolDefinition.path.exists
-      destinationDoc = bestTextDocument(symbolDefinition)
-      destinationPath = symbolDefinition.path.toFileOnDisk(workspace)
-      destinationDistance =
+  def definition(
+      symbol: String,
+      source: Option[AbsolutePath]
+  ): Option[SymbolDefinition] = {
+    val allowedBuildTargets =
+      for {
+        src <- source
+        id <- buildTargets.inverseSources(src)
+      } yield buildTargets.buildTargetTransitiveDependencies(id)
+
+    val definitions = index.definitions(Symbol(symbol)).filter(_.path.exists)
+    allowedBuildTargets match {
+      case None => definitions.headOption
+      case Some(targets) =>
+        val asSet = targets.toSet
+        val matched = definitions.find { defn =>
+          sourceBuildTargets(defn.path).exists(id => asSet.contains(id))
+        }
+        // Fallback to any definition - it's needed for worksheets
+        // They might have dynamic `import $dep` and these sources jars
+        // aren't registered in buildTargets
+        matched.orElse(definitions.headOption)
+    }
+  }
+
+  private def sourceBuildTargets(
+      source: AbsolutePath
+  ): List[BuildTargetIdentifier] = {
+    def trivialSource: Option[List[BuildTargetIdentifier]] =
+      Option(buildTargets.sourceBuildTargets(source).toList).filter(_.nonEmpty)
```

Comment:
Couldn't this be simplified to just `List`? Whenever I see `Option[List]` I always think that unless there is a real difference between `None` and `Some(List.empty)`.

Reply:
Yep, I agree that in most cases `Option[List[A]]` should be reduced to `List[A]`.

However, here it's used only in inner functions just to take benefits from `orElse` combinator in usage below - https://github.com/scalameta/metals/blob/a82a14240bd55284ba220650fe5a70c4198b5cde/metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala#L305-L308
Otherwise, I had to inекщвгсу smth like `orElseIfNil` for `List` for this small part of code.

Reply:
Alright sounds good 👍🏼